src/track/external-captions-handler.js

```diff
@@ -0,0 +1,397 @@
+    this._eventManager.listen(this._player, this._player.Event.TEXT_TRACK_CHANGED, (e) => {
+      this._handleTextTrackChanged(e)
+    });
+  }
+
+  /**
+   * Make a request to download a caption and parse it's content to cues.
+   * @param {TextTrack} textTrack - download and parse the cues of the text track
+   * @returns {Promise<any>} - resolves when the request returns and the caption string is parsed to cues.
+   * @private
+   */
+  _getCuesArray(textTrack: TextTrack): Promise<*> {
+    let cuesPromise = new Promise((resolve, reject) => {
+      Utils.Http.execute(textTrack.url, {}, 'GET').then(response => {
+        if ((typeof textTrack.captionsType != 'undefined' && textTrack.captionsType === ExternalCaptionsHandler.SRT_POSTFIX) || this._getFileType(textTrack.url) === ExternalCaptionsHandler.SRT_POSTFIX) {
+          response = this._convertSrtToVtt(response);
+        }
+        this._parseCues(response, resolve);
+      }).catch(error => {
+        reject(new Error(Error.Severity.RECOVERABLE, Error.Category.TEXT, Error.Code.HTTP_ERROR, error.payload));
+      })
+    });
+    this._cuesInProcessMap[textTrack.language] = cuesPromise;
+    return cuesPromise;
+  }
+
+  /**
+   * this calls the VTTCue parser and parse the .vtt captions string into vttCues objects
+   * @param {string} vttStr - a string in a .vtt format to parse into VTTCues
+   * @param {Function} resolve - a resolve function of the calling function.
+   * @returns {void}
+   * @private
+   */
+  _parseCues(vttStr: string, resolve: Function): void {
+    const parser = new Parser(window, StringDecoder());
+    let cues = [];
+    parser.oncue = cue => {
+      cues.push(cue);
+    };
+    parser.onflush = () => {
+      ExternalCaptionsHandler._logger.debug('finished parsing external cues');
+      resolve(cues);
+    };
+    parser.parse(vttStr);
+    parser.flush();
+  }
+
+  /**
+   * converts a .SRT string into .VTT string
+   * @param {string} str - a string in a .SRT format
+   * @returns {string} - a string in a .VTT format
+   * @private
+   */
+  _convertSrtToVtt(str: string): string {
+    return str.replace(/(\d\d:\d\d:\d\d).(\d\d\d) --> (\d\d:\d\d:\d\d).(\d\d\d)/g, (match, part1, part2, part3, part4) => {
+      return `${part1},${part2} --> ${part3},${part4}`;
+    });
+  }
+
+  /**
+   * resolves with a caption object that contains all the caption data
+   * start the parsing, creation and addition of the external captions.
+   * @param {TextTrack} textTrack - create a single caption. when the process ends, the caption is removed from the
+   * _cuesInProcessMap.
+   * @returns {Promise<any>} - the caption object with the parsed cues
+   * @private
+   */
+  _createCaption(textTrack: TextTrack): Promise<*> {
+    return new Promise((resolve) => {
+      this._getCuesArray(textTrack)
+        .then(cues => {
+          textTrack.cues = cues;
+          this._player.setTextTrackCues(textTrack);
+          delete this._cuesInProcessMap[textTrack.language];
```

Comment:
better manage this as a consistent map of tracks to states.
Instead of deleting it just create a tracks in process (it's not the cues that are in process, but the entire track being downloaded and parsed).
set a state map of `NOT_DOWNLOADED`, `DOWNLOADING` and `READY` and set it in each state.